SL pbi beta (#7000)
fishtown-analytics/docs.getdbt.com: website/docs/docs/cloud-integrations/semantic-layer/power-bi.md

   <Lightbox src="/img/docs/cloud-integrations/sl-pbi/pbi-encrypted.jpg" title="Set the connection as Encrypted" />
6. Click **Create**. This will run a connection test (unless you choose to skip it). If the connection succeeds, the connection will be saved.

You can now go back to your published report on Power BI Service to assert data loads as expected.

## Use the connector

This section describes how to use the dbt Semantic Layer connector in Power BI.

The dbt Semantic Layer connector creates:
- A virtual (or sample?) table for each saved query
- A `METRICS.ALL` table containing all metrics, and dimensions and entities appear as regular dimension columns.

These tables do not actually map to an underlying table in your data warehouse. Instead, Power BI sends queries to these tables and (before actually executing on the warehouse) the Semantic Layer servers:
- Parse the SQL
- Extract all the queried columns, group bys and filters 
- Generates SQL to query your existing tables. 
- Returns data back to Power BI, which doesn’t know any of this happened.

> ADD DIAGRAM

Author: waiting for [design diagram](https://app.asana.com/1/312418853397011/project/1202263237307725/task/1209568430873549?focus=true)

---
title: "Power BI"
description: "Use Power BI to query the dbt Semantic Layer and produce dashboards with trusted data."
tags: [Semantic Layer]
sidebar_label: "Power BI"
---

# Power BI <Lifecycle status="enterprise,beta" />

The Power BI integration enables you to query the dbt Semantic Layer directly, allowing you to build dashboards with trusted, live data in Power BI. It provides a live connection to the dbt Semantic Layer through Power BI Desktop or Power BI Service.

:::tip
The Power BI integration is currently in private beta. If you're interested in participating, please reach out to your account representative.
:::

> ## Prerequisites

Author: @serramatutu have added the diagram!

- You have [configured the dbt Semantic Layer](/docs/use-dbt-semantic-layer/setup-sl) 
- You are on a supported [dbt Cloud release track](/docs/dbt-versions/cloud-release-tracks) or on dbt v1.6 or higher.

> - You installed [Power BI Desktop or Power BI On-premises Data Gateway](https://learn.microsoft.com/en-us/power-bi/connect-data/service-gateway-custom-connectors)

Reviewer: ```suggestion
- You installed [Power BI Desktop or Power BI On-premises Data Gateway](https://learn.microsoft.com/en-us/power-bi/connect-data/service-gateway-custom-connectors).
```

## For IT admins

This section is for IT admins trying to install the ODBC driver and connector into an On-premises Data Gateway.

To allow published reports to use the connector in Power BI Service, an IT Admin must install and configure the connector.

#### 1. Install the ODBC driver and connector into an On-premises Data Gateway
   - Run the same `.msi` installer used for Power BI Desktop.
   - Install it on the machine where your gateway is hosted.

#### 2. Copy connector file to Gateway directory
   - Locate that `.pqx` file: `C:\Users\<YourUser>\Documents\Power BI Desktop\Custom Connectors\dbtSemanticLayer.pqx`.
   - Copy it to the Power BI On-premises Data Gateway custom connectors directory: `C:\Windows\ServiceProfiles\PBIEgwService\Documents\Power BI Desktop\Custom Connectors`.
#### 3. Verify installation
   - Following verification steps from the [install the connector](#3-verify-installation) section.
#### 4. Enable connector in Power BI Enterprise Gateway
   - Open the `EnterpriseGatewayConfigurator.exe`.

>    - Navigate to **Connectors** 

Reviewer: ```suggestion
   - Navigate to **Connectors**. 
```

   - Verify that the `dbt Semantic Layer` connector is installed and active.

For more information on how to set up custom connectors in the Power BI On-premises Data Gateway, refer to Power BI’s [official documentation](https://learn.microsoft.com/en-us/power-bi/connect-data/service-gateway-custom-connectors).

## Configure the connector
After installing the connector, you’ll have to configure your project credentials to connect to the Semantic Layer from a report.

To configure project credentials in Power BI Desktop:

1. Create a blank report.
2. On the top-left, click on **Get data**.
3. Search for dbt Semantic Layer, then click **Connect**.

> 4. Fill in your connection details. You can find your server and environmentId under the Semantic Layer configuration for your dbt Cloud project.

Reviewer: ```suggestion
4. Fill in your connection details. You can find your server and environment ID under the Semantic Layer configuration for your dbt Cloud project.
```

4. Fill in your connection details. You can find your server and environment ID under the Semantic Layer configuration for your dbt Cloud project.
   :::tip 
   Make sure you select `DirectQuery`, since the Semantic Layer connector does not support `Import` mode (see [Considerations](#considerations) for more details). 
   :::
5. Click **OK** to proceed.
   <Lightbox src="/img/docs/cloud-integrations/sl-pbi/pbi-directquery.jpg" title="Select DirectQuery mode" />
6. On the next screen, paste your service token and then click **Connect**.

> 7. You should see a sidepane with a few tables. Select the ones you want to load into your dashboard. Then click **Load**.

Reviewer 2: ```suggestion
7. You should see a sidepane with a few "virtual" tables. `ALL` represents all of your defined SL objects, and the other tables represent each of your Saved Queries. Select the one you want to load into your dashboard. Then click **Load**.
```

   - If the Status is **Not configured correctly**, you’ll have to configure it.
   <Lightbox src="/img/docs/cloud-integrations/sl-pbi/pbi-gateway-cloud-connections.jpg" title="Configure the gateway connection" />
4. Click on the arrow under **Actions** and then, click on **Manually add to gateway**.
   <Lightbox src="/img/docs/cloud-integrations/sl-pbi/pbi-manual-gateway.jpg" title="Manually add to gateway" />
5. Provide a name for your connection and enter your connection details. 
   - Set the connection as **Encrypted** (Required). Failing to do so will result in the Semantic Layer servers rejecting the connection.
   <Lightbox src="/img/docs/cloud-integrations/sl-pbi/pbi-encrypted.jpg" title="Set the connection as Encrypted" />
6. Click **Create**. This will run a connection test (unless you choose to skip it). If the connection succeeds, the connection will be saved.

You can now go back to your published report on Power BI Service to assert data loads as expected.

## Use the connector

This section describes how to use the dbt Semantic Layer connector in Power BI.

The dbt Semantic Layer connector creates:

> - A virtual table for each saved query

Reviewer: ```suggestion
- A virtual table for each saved query.
```

- A `METRICS.ALL` table containing all metrics, and dimensions and entities appear as regular dimension columns.

These tables do not actually map to an underlying table in your data warehouse. Instead, Power BI sends queries to these tables and (before actually executing on the warehouse) the Semantic Layer servers:

> - Parse the SQL

Reviewer: ```suggestion
- Parse the SQL.
```

> - Extract all the queried columns, group bys and filters 

Reviewer: ```suggestion
- Extract all the queried columns, group bys and filters.
```

- Generates SQL to query your existing tables. 
- Returns data back to Power BI, which doesn’t know any of this happened.

ADD DIAGRAM

> This allows for very flexible analytics workflows, like drag and drop metrics and slice by dimensions and entities &mdash; the Semantic Layer will generate the appropriate SQL to actually query your data source for you.

Reviewer: ```suggestion
This allows for very flexible analytics workflows, like drag and drop metrics, slice by dimensions and entities &mdash; the Semantic Layer will generate the appropriate SQL to actually query your data source for you.
```

> ## Considerations

Reviewer 2: We'll want to add another consideration:

- Custom modeling is not supported
  - Joining tables is not supported
  - Creating custom columns within a table is not supported
  - Custom DAX or PQ is not supported (although some of it might work)

- To import data into a PowerBI report:
   - Select a valid combination of columns to import, (something that will generate a valid Semantic Layer query)
   - You can use `Table.SelectColumns` for this: `= Table.SelectColumns(Source{[Item="ALL",Schema="METRICS",Catalog=null]}[Data], {"Total Profit", "Metric Time (Day)"})`
   - Be aware that all calculations will happen inside of Power BI and won’t pass through Semantic Layer servers. This could lead to incorrect or diverging results.
   - For example, the Semantic Layer is usually responsible for rolling up cumulative metrics to coarser time granularities. Doing a sum over all the weeks in a year to get a yearly granularity out of a weekly Semantic Layer query will most likely generate incorrect results. Instead, you should query the Semantic Layer directly to get accurate results.

</Expandable>

<Expandable alt_header="The dbt Semantic Layer connector ignores aggregations defined in Power BI">

> - If you change the aggregation type of a metric from `SUM()` to `COUNT()` or anything else, nothing will change. This is because aggregation functions are defined in the Semantic Layer and we ignore them when translating Power BI generated SQL into Semantic Layer queries.

Reviewer 2: We'll want to add another note mentioning that:
- `Count (Distinct)`, `Standard Deviation`, `Variance`, and `Median` might all throw an error and not work at all

Author: thank you @aiguofer !

>- Aggregations like `Count (Distinct)`, `Standard Deviation`, `Variance`, and `Median` in Power BI may return an error and not work at all.

---
title: "Power BI"
description: "Use Power BI to query the dbt Semantic Layer and produce dashboards with trusted data."
tags: [Semantic Layer]
sidebar_label: "Power BI"
---

# Power BI <Lifecycle status="enterprise,beta" />

The Power BI integration enables you to query the dbt Semantic Layer directly, allowing you to build dashboards with trusted, live data in Power BI. It provides a live connection to the dbt Semantic Layer through Power BI Desktop or Power BI Service.

:::tip

> The Power BI integration is currently in private beta. If you're interested in participating, please reach out to your account representative.

Author: ```suggestion
The dbt Semantic Layer Power BI integration is currently in private beta. To join the beta, contact your account representative.
```
